### ISSUE TITLE
Interaction between CUDAGraphs and AttentionBias

### ISSUE THREAD

**Issue by TeodorPoncu** [facebookresearch/xformers#968]
# ❓ Questions and Help

Recently I've been trying to use `torch.cuda.CUDAGraphs` in combination with `BlockDiagonalCausalMask` in order to get better training speeds.

I've mostly taken inspiration from this script: https://github.com/facebookresearch/xformers/blob/main/examples/llama_inference/generate.py 

For some sequence lengths (up to 256 tokens), I was able to get equivalent results to the eager mode, however when going over a specific sequence size, the results are no longer equivalent.

 My question is: is there a known / stable / safe way of varying the sequence lengths when using `torch.cuda.CUDAGraphs`?

I've included the code that I'm using to test down bellow. 
```python
from typing import List
import xformers.ops as xops
from xformers.ops.fmha.attn_bias import (
  AttentionBias,
  BlockDiagonalCausalMask,
)

import torch
import xformers
from torch import Tensor

class Attention(torch.nn.Module):
  def __init__(self, *args, **kwargs) -> None:
    super().__init__(*args, **kwargs)

  def forward(
    self, q: Tensor, k: Tensor, v: Tensor, mask: AttentionBias) -> Tensor:
    return xops.memory_efficient_attention(
      q, k, v, mask
    )

def eager_from_seqlens(fn: Attention, q: Tensor, k: Tensor, v: Tensor, seqlens: List[int]):
  o = fn.forward(q=q, k=k, v=v, mask=BlockDiagonalCausalMask.from_seqlens(seqlens))
  return o

def eager_with_mask(fn: Attention, q: Tensor, k: Tensor, v: Tensor, mask: AttentionBias,):
  o = fn.forward(q=q, k=k, v=v, mask=mask)
  return o

def set_mask_seqlens(mask: AttentionBias, seqlens: List[int]):
  min_seqlen = min(seqlens)
  max_seqlen = max(seqlens)
  
  new_seqstart = torch.zeros_like(mask.q_seqinfo.seqstart)
  len_seqstart = len(mask.q_seqinfo.seqstart)
  
  seqlen_padded = seqlens + [0] * (len_seqstart - len(seqlens) - 1)
  cum_sum_seqlens = torch.cumsum(torch.tensor(seqlen_padded, dtype=torch.long), dim=0)


  new_seqstart[1:len(seqlen_padded)+1] = cum_sum_seqlens
  new_seqstart_py = new_seqstart.tolist()   

  mask.q_seqinfo.seqstart.copy_(new_seqstart)
  mask.q_seqinfo.seqstart_py = new_seqstart_py
  mask.q_seqinfo.seqlens_py = seqlens
  
  mask.q_seqinfo.max_seqlen = max_seqlen
  mask.q_seqinfo.min_seqlen = min_seqlen

def graph_from_seqlens(fn: Attention, q: Tensor, k: Tensor, v: Tensor, seqlens: List[int]):
  mask = BlockDiagonalCausalMask.from_seqlens([0] * 512)
  mask.q_seqinfo.to(device=torch.device("cuda"))
  mask.k_seqinfo.to(device=torch.device("cuda"))
  set_mask_seqlens(mask, seqlens)

  graph = torch.cuda.CUDAGraph()
  for _ in range(3):
    o = fn.forward(x, x, x, mask)

  with torch.cuda.graph(graph):
    o = fn.forward(x, x, x, mask)

  graph.replay()
  torch.cuda.synchronize()
  return o

def graph_from_subsequent_seqlens(fn: Attention, q: Tensor, k: Tensor, v: Tensor, seqlens: List[int],):
  mask = BlockDiagonalCausalMask.from_seqlens([0] * MAX_MASK_LEN)
  mask.q_seqinfo.to(device=torch.device("cuda"))
  mask.k_seqinfo.to(device=torch.device("cuda"))

  # run some dummy mask
  set_mask_seqlens(mask, [2] * (MAX_MASK_LEN // 2)) 

  graph = torch.cuda.CUDAGraph()
  for _ in range(3):
    o = fn.forward(x, x, x, mask)

  with torch.cuda.graph(graph):
    o = fn.forward(x, x, x, mask)

  # dummy forward
  graph.replay()
  torch.cuda.synchronize()

  # actual iteration with an updated mask
  # meant to simulate training behaviour when seqlens would have
  # varrying lengths due to how samples get mixed in the sequence
  set_mask_seqlens(mask, seqlens)
  graph.replay()
  torch.cuda.synchronize()

  return o

if __name__ == "__main__":
  print(torch.__version__)
  print(xformers.__version__)

  MAX_MASK_LEN = 256
  HALF_MASK_LEN = MAX_MASK_LEN // 2

  torch.set_default_device("cuda")
  x = torch.randn(1, MAX_MASK_LEN, 8, 32)
  x = x.half()

  attn = Attention()

  # normal seqlens list
  o1 = eager_from_seqlens(fn=attn, q=x, k=x, v=x,
                          seqlens=[HALF_MASK_LEN, HALF_MASK_LEN])

  # seqlens list with 0 padding  
  o2 = eager_from_seqlens(fn=attn, q=x, k=x, v=x,
                          seqlens=[HALF_MASK_LEN, HALF_MASK_LEN] + [0] * (MAX_MASK_LEN - 2))


  # mask with edited seqinfo tensors
  fixed_seqlens = [0] * MAX_MASK_LEN
  
  # we instantiate the mask with a fixed seqlen list
  # in order to generate a fixed size tensor to which we can copy
  # various zero padded tensor
  mask = BlockDiagonalCausalMask.from_seqlens(fixed_seqlens)
  set_mask_seqlens(mask, [HALF_MASK_LEN, HALF_MASK_LEN])

  o3 = eager_with_mask(attn=attn, q=x, k=x, v=x,
                       mask=mask)
  
  # graph mode with 1 single replay
  o4 = graph_from_seqlens(attn=attn, q=x, k=x, v=x,
                          seqlens=[HALF_MASK_LEN, HALF_MASK_LEN])
  
  # graph mode with where the last graph replay uses the provided seqlens
  o5 = graph_from_subsequent_seqlens(attn=attn, q=x, k=x, v=x,
                                     seqlens=[HALF_MASK_LEN, HALF_MASK_LEN])

  # eager mode embedding
  print(o1[0][-1][0])

  # graph mode with mask edditing
  print(o5[0][-1][0])

  assert torch.allclose(o1, o2, rtol = 1e-8)
  assert torch.allclose(o1, o3, rtol = 1e-8)
  assert torch.allclose(o1, o4, rtol = 1e-8)
  assert torch.allclose(o1, o5, rtol = 1e-8)
```

Setting `MAX_MASK_LEN = 512` will fail the last assert on my environment, and the results will also be wildly different (so not an `rtol` issue).

My torch version is: `2.3.0.dev20240116+cu121`
My xformers version is: `0.0.24+f540d2b.d20240123`

**Comment by bottler** [comment]
The general rule is that changes to values on the CPU will not be seen when you replay a graph. 

In this case, I think the only problem is that the value of `attn_mask.q_seqinfo.max_seqlen` (an int) which is also `attn_mask.k_seqinfo.max_seqlen` will have been baked into the graph (and it probably determines the launch configuration of a kernel). I think everything will still work if this number is set too high when you calculate attention, but the calculation will be wrong if it is too low. So you could edit it manually to the maximum of any seqlen you might use, straight after you construct `mask`. For safety, instead of setting max_seqlen and min_seqlen in set_mask_seqlens, you could just assert that the new values are between the existing min and max .

**Comment by TeodorPoncu** [comment]
Thank you very much @bottler! Setting a static absolute min and absolute max solved my issue. 

I would've definitely not thought about that as for some reason I was under the impression that `q_seqinfo.max_seqlen / q_seqinfo.min_seqlen` had to be dynamically tied to the minimum and maximum value provided in the `seqlens` list.
